docs: Docker Engine on Windows how-to article (#4406)
FlowFuse/flowfuse: docs/install/docker/windows-docker-ce.md

---
navTitle: Docker Engine on Windows
meta:
   description: Install Docker Engine (Docker CE) on Windows.

> ---

Reviewer: Adding long-tail keywords will be beneficial. I have suggested some, but feel free to update them as needed.

```suggestion
   tags: 
       - docker for windows
       - install docker on windows
       - install docker on windows with wsl2
       - running docker on windows 11
---
```

# How to Install Docker Engine (Docker CE) on Windows using WSL2

## Disclaimer 

> Docker Containers are a Linux technology, and the setup described here using Docker on Windows via WSL2 should not be considered production-ready. It is more suitable for development, testing, or learning purposes.

Reviewer: ```suggestion
Docker containers are inherently a Linux technology. The setup described here for running Docker on Windows using WSL2 is intended for development, testing, or learning purposes and may not be suitable for production environments.
```

## Prerequisites

> Ensure your system meets following requirements:

Reviewer: ```suggestion
Ensure your system meets the following requirements:
```

- Windows Server 2019 (version 1709) and later
- Windows 10 version 2004 and higher (Build 19041 and higher)
- Windows 11

## Step 1: Install Windows Subsystem for Linux 

Open PowerShell as an administrator and run the following commands:

1. Install WSL and default Linux distribution (Ubuntu) using the following command:
   ```powershell
   wsl --install
   ```

2. Reboot your system to apply changes
   ```powershell
   shutdown -r -t 5
   ```

   After the reboot, WSL will automatically start installing Ubuntu Linux distribution.
   You will be prompted to create a new UNIX account. Follow the instructions to create a new user account and set a password.
   ![wsl-user](../images/wsl-unix-user.png)

   Once completed, you will be dropped into a new Ubuntu shell.

>    ![wsl-complete](../images/wsl-install-complete.png)

Reviewer: Hi @ppawlowski, could you please update the alt text to include more detailed information?

     $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
     sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
   sudo apt-get update
   ```

3. **Install Docker Packages**

   Now, install Docker and its associated packages:

   ```bash
   sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
   ```

4. **Start the Docker Service**

   Start Docker using the following command:

   ```bash
   sudo /etc/init.d/docker start
   ```

5. **Verify Docker Installation**

   Run a test Docker container to verify that Docker is installed and running correctly:

   ```bash
   sudo docker run hello-world
   ```

   If Docker is installed correctly, you should see the similar output:

   ![wsl-docker-complete](../images/wsl-docker-complete.png)

Once Docker is installed, you can [install FlowFuse platform using docker compose](./README.md).

## Optional: Port Forwarding from External IP Address to Docker Container

If you want to forward a port from your external IP address to a Docker container, follow these additional steps.

Consider a scenario where you want to run an Nginx container:

1. **Run the Nginx Container**

>    In Linux shell, start the Nginx container, binding port 8080 on your localhost to port 80 in the container:

Reviewer: ```suggestion
   In the Linux shell, start the Nginx container, binding port 8080 on your localhost to port 80 in the container:
```

   ```bash
   sudo docker run -d -p 8080:80 --name mynginx nginx
   ```

2. **Set Up Port Forwarding**

   To forward traffic from an external IP to your container, run the following PowerShell command (administrator privileges required):

   ```powershell
   netsh interface portproxy add v4tov4 listenport=8080 listenaddress=0.0.0.0 connectport=8080 connectaddress=127.0.0.1
   ```
   

>    This command forwards traffic from port 8080 on your external IP address to port 8080 on your localhost, where the Nginx container is listening for a connections.

Reviewer: ```suggestion
   This command forwards traffic from port 8080 on your external IP address to port 8080 on your localhost, where the Nginx container is listening for connections.
```

---
navTitle: Docker Engine on Windows
meta:
   description: Install Docker Engine (Docker CE) on Windows.
   tags: 
       - docker for windows
       - install docker on windows
       - install docker on windows with wsl2
       - running docker on windows 11
---

# How to Install Docker Engine (Docker CE) on Windows using WSL2

## Disclaimer 

Docker containers are inherently a Linux technology. The setup described here for running Docker on Windows using WSL2 is intended for development, testing, or learning purposes and may not be suitable for production environments.

## Prerequisites

Ensure your system meets the following requirements:

> - Windows Server 2019 (version 1709) and later

Reviewer 2: @ppawlowski 

WSL2 is not available on Server 2019 and non updated systems (only WSL)

Rather than listing out all the ways it can be installed, i would recommend a generic method and min OS version + a quick check.

e.g ask the user to run `wsl --status` to confirm that WSL2 is being used.

```
❯ wsl --status
Default Distribution: Ubuntu
Default Version: 2
```

---
navTitle: Docker Engine on Windows
meta:
   description: Install Docker Engine (Docker CE) on Windows.
   tags: 
       - docker for windows
       - install docker on windows
       - install docker on windows with wsl2
       - running docker on windows 11
---

# How to Install Docker Engine (Docker CE) on Windows using WSL2

> ## Disclaimer 
> 
> Docker containers are inherently a Linux technology. The setup described here for running Docker on Windows using WSL2 is intended for development, testing, or learning purposes and may not be suitable for production environments.

Reviewer 3: ```suggestion
We recommend running FlowFuse Docker on Linux when ever possible as containers are inherently a Linux technology. If that really is not possible in your environment and the base OS must be Windows then the next best option is to use the WSL2 (Windows Subsystem for Linux v2) feature of Windows.
```

## Prerequisites

> Ensure your system meets the following Windows Subsustem form Linux 2 requirements:

Reviewer 3: ```suggestion
Ensure your system meets the following Windows Subsystem for Linux v2 requirements:
```
